improve & test ServerName debug output - workaround

brodybits opened the pull request:
**hopefully** resolves #1177

**before:**

result from `format!("{:?}", rustls::ServerName::try_from("a.com"))`:

    Ok(DnsName(DnsName(DnsName(\"a.com\"))))

**after:**

result from `format!("{:?}", rustls::ServerName::try_from("a.com"))`:

    Ok(DnsName(\"a.com\"))

**also tested:**

`rustls::ServerName::IpAddress("127.0.0.1".parse().unwrap())` formats as:

    IpAddress(127.0.0.1)

As I said in 3a3e203d51f08463719513f0bd6cb5f7716631cb, I think the root cause lies in the difference between the standard debug formatting of tuples vs enums in Rust.

Quick disclaimer that I just started learning Rust after many years of programming and this is actually my first ever contribution in this language. I made a simple workaround solution for now but think it would be ideal to consider further improvement in the future. Please feel free to leave **candid** feedback or even rework before merging if needed.

I would personally favor merging this one in a squash commit and would be happy to squash the changes myself if needed.

djc commented on rustls/src/verify.rs:
```diff
@@ -212,6 +212,15 @@ impl AsRef<str> for DnsName {
 }
 
+impl fmt::Debug for DnsName {
+    // Workaroud solution for ServerName debug formatting:
```
Tiny nit: "workaroud" -> "workaround".

brodybits replied:
> Tiny nit: "workaroud" -> "workaround".

making this from the GH GUI:

```suggestion
    // Workaround solution for ServerName debug formatting:
```

djc reviewed (approved):
Thank you for the contribution, and welcome to the Rust community!

This looks like a good change, if you want to squash the commits and leave one commit with a nice commit message that would be great -- otherwise we can squash during merge.

brodybits commented on rustls/tests/api.rs:
```diff
@@ -4611,3 +4611,22 @@ fn test_received_plaintext_backpressure() {
+#[test]
+fn test_debug_server_name_from_string() {
+    assert_eq!(
+        format!("{:?}", rustls::ServerName::try_from("a.com")),
+        "Ok(DnsName(\"a.com\"))"
```
```suggestion
        format!("{:?}", rustls::ServerName::try_from("a.com").unwrap()),
        "DnsName(\"a.com\")"
```

brodybits commented:
Thanks @djc for the quick review!

You should see that I have applied the nit, unwrapped a test result, and rebased (twice) with a cleaner commit message. I have also updated the title of this PR.

I have also started working on a slightly larger proposal in PR #1193, which may be a little controversial and needs more work to get the debug output right. I would probably favor keeping PR #1193 in a separate stage, though may come with a little more code churn.

I gotta say that after years in some major systems programming I struggled for a while to grasp the most important concepts of Rust, eventually found this helpful: https://stevedonovan.github.io/rust-gentle-intro/
